File: codes/01_advDiffSolver/program.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import numpy as np
from time import time
import matplotlib.pyplot as plt

nHalo = 2
sIn = slice(nHalo, -nHalo)
def updateHalo(u):
    nH2 = 2*nHalo
    u[0:nHalo, sIn] = u[-nH2:-nHalo, sIn]
    u[-nHalo:, sIn] = u[nHalo:2*nHalo, sIn]
    u[sIn, 0:nHalo] = u[sIn, -nH2:-nHalo]
    u[sIn, -nHalo:] = u[sIn, nHalo:2*nHalo]


cAdv = np.array([ 1./12, -2./3,  0,    2./3, -1./12])
cDif = np.array([-1./12,  4./3, -5./2, 4./3, -1./12])

class Problem:

    def __init__(self, fileName):
        with open(fileName, "r") as f:
            inputs = f.read().split()

        self.nX, self.nY = [int(inputs.pop(0)) for _ in range(2)]
        self.initType, self.flowType = [inputs.pop(0) for _ in range(2)]
        self.viscosity, self.tEnd = [float(inputs.pop(0)) for _ in range(2)]
        self.nSteps = int(inputs.pop(0))

        self.setupSolution()
        self.setupCoeffs()

        # Additional variable
        self.t = 0
        self.tmp = np.empty((self.nX, self.nY))

    @property
    def grid(self):
        x = np.linspace(0, 1, self.nX, endpoint=False)[:, None]
        y = np.linspace(0, 1, self.nY, endpoint=False)[None, :]
        return x, y


    def setupSolution(self):
        self.u = u = np.zeros((self.nX+2*nHalo, self.nY+2*nHalo))
        initType, (x, y) = self.initType, self.grid

        if initType == "gauss":
            u[sIn, sIn] = np.exp(-200*((x-0.5)**2 + (y-0.5)**2))
        elif initType == "square":
            u[sIn, sIn] = (x > 0.2)*(x < 0.3)*(y > 0.2)*(y < 0.3)
        elif initType == "cross":
            u[sIn, sIn] = 0.5*(np.exp(-200*(x-0.5)**2) + np.exp(-200*(y-0.5)**2))
        elif initType == "cross2":
            u[sIn, sIn] = np.maximum(np.exp(-200*(x-0.5)**2), np.exp(-200*(y-0.5)**2))
        else:
            raise ValueError(f"unknown initType : {initType}")


    def setupCoeffs(self):
        self.coeffs = coeffs = np.zeros((2, 2*nHalo+1, self.nX, self.nY))

        flowType, viscosity = self.flowType, self.viscosity
        dX, dY, (x, y) = 1/self.nX, 1/self.nY, self.grid

        if flowType == "diagonal":
            vX, vY = 1, 1
        elif flowType == "circular":
            r = np.hypot(x-0.5, y-0.5)
            phi= np.arctan2(y-0.5, x-0.5)
            rho = np.exp(-10*r**2)
            vX = -r*2*np.pi*np.sin(phi)*rho
            vY =  r*2*np.pi*np.cos(phi)*rho
        elif flowType == "circular2":
            r = np.hypot(x-0.5, y-0.5)
            phi= np.arctan2(y-0.5, x-0.5)
            rho = np.exp(-5*r**2)
            vX = -r*2*np.pi*np.sin(phi)*np.sin(4*np.pi*r)*rho
            vY =  r*2*np.pi*np.cos(phi)*np.sin(4*np.pi*r)*rho
        else:
            vX = vY = 0
        coeffs[0] = -vX*cAdv[:, None, None]/dX + viscosity*cDif[:, None, None]/dX**2
        coeffs[1] = -vY*cAdv[:, None, None]/dY + viscosity*cDif[:, None, None]/dY**2


    def computeRHS(self, uEval, t, out):
        coeffs, tmp, nX, nY = self.coeffs, self.tmp, self.nX, self.nY

        updateHalo(uEval)

        out[:] = 0
        for s in range(2*nHalo+1):

            # Derivative in X
            np.copyto(tmp, uEval[s:nX+s, sIn])
            tmp *= coeffs[0, s]
            out += tmp

            # Derivative in Y
            np.copyto(tmp, uEval[sIn, s:nY+s])
            tmp *= coeffs[1, s]
            out += tmp


    def simulate(self):
        u0, nX, nY = self.u, self.nX, self.nY
        uEval = np.zeros_like(u0)

        u1 = np.empty((nX, nY))
        np.copyto(u1, u0[sIn, sIn])
        k = np.zeros_like(u1)

        dt = self.tEnd/self.nSteps
        tBeg = time()
        for i in range(self.nSteps):
            t = self.t

            self.computeRHS(u0, t, k)
            np.copyto(uEval[sIn, sIn], k)
            k *= dt/6; u1 += k

            uEval *= dt/2
            uEval += u0
            self.computeRHS(uEval, t+dt/2, k)
            np.copyto(uEval[sIn, sIn], k)
            k *= dt/3; u1 += k

            uEval *= dt/2
            uEval += u0
            self.computeRHS(uEval, t+dt/2, k)
            np.copyto(uEval[sIn, sIn], k)
            k *= dt/3; u1 += k

            uEval *= dt
            uEval += u0
            self.computeRHS(uEval, t+dt, k)
            k *= dt/6; u1 += k

            np.copyto(u0[sIn, sIn], u1)
            self.t = (i+1)*dt

        tWall = time()-tBeg
        print(f"tWall : {tWall}")
        print(f"tWall/DoF : {tWall/(self.nSteps*nX*nY)}")


# Simulation
p = Problem("input.txt")
u0 = p.u.copy()

p.simulate()
uEnd = np.abs(p.u)  # remove some negative artefacts for plots

# Plotting
fig, ax = plt.subplots(1, 2, figsize=(12, 5))

grid = [x.ravel() for x in p.grid]

c0 = ax[0].contourf(*grid, u0[sIn, sIn])
ax[0].set_title('T=0')
ax[0].set_xlabel('X')
ax[0].set_ylabel('Y')
fig.colorbar(c0, ax=ax[0], orientation='vertical')

c1 = ax[1].contourf(*grid, uEnd[sIn, sIn])
ax[1].set_title(f'T={p.tEnd}')
ax[1].set_xlabel('X')
ax[1].set_ylabel('Y')
fig.colorbar(c1, ax=ax[1], orientation='vertical')

plt.tight_layout()
plt.savefig("solution.png")

# plt.show()

File: codes/01_advDiffSolver/src/program.jl
## import
#using LinearAlgebra
using BenchmarkTools
using Plots; gr()

## define
@Base.kwdef mutable struct adv_diff_parameters
    # Inputs from file with defaults
    nX        :: Int                = 256 
    nY        :: Int                = 256
    initType  :: String             = "cross2"
    flowType  :: String             = "circular2"
    viscosity :: Float64            = 0.0
    tEnd      :: Float64            = 0.125
    nSteps    :: Int                = 125
    # Other
    dlo       :: Tuple{Int, Int}    = (nHalo+1, nHalo+1)
    dhi       :: Tuple{Int, Int}    = (nHalo+nX, nHalo+nY)
    ndof      :: Tuple{Int, Int}    = (2*nHalo+nX, 2*nHalo+nY)
end # struct adv_diff_parameters

function read_inputs_and_update_parameters!(filename::String,parms::adv_diff_parameters)
    f = open(filename, "r")
    lines = readlines(f)
    parms.nX, parms.nY = parse.(Int,split(lines[1]))
    line = split(lines[2])
    parms.initType, parms.flowType = line[1:2]
    parms.viscosity = parse(Float64, line[3])
    line = split(lines[3])
    parms.tEnd = parse(Float64,line[1])
    parms.nSteps = parse(Int,line[2])
    # update derived parameters
    parms.dlo = (nHalo+1, nHalo+1)
    parms.dhi = (nHalo+parms.nX, nHalo+parms.nY)
    parms.ndof = (2*nHalo+parms.nX, 2*nHalo+parms.nY)
    # close file
    close(f)
end # read_inputs_and_update_parameters!

function updateHalo!(u::Matrix{Float64})
    nH2 = 2*nHalo
    u[1:nHalo,         nHalo+1:end-nHalo] = u[end-nH2+1:end-nHalo, nHalo+1:end-nHalo]
    u[end-nHalo+1:end, nHalo+1:end-nHalo] = u[nHalo+1:nH2,         nHalo+1:end-nHalo]
    u[nHalo+1:end-nHalo,         1:nHalo] = u[nHalo+1:end-nHalo, end-nH2+1:end-nHalo]
    u[nHalo+1:end-nHalo, end-nHalo+1:end] = u[nHalo+1:end-nHalo,         nHalo+1:nH2] 
end # updateHalo!

function generate_grid(parms::adv_diff_parameters)
    Δx, Δy = 1.0/parms.nX, 1.0/parms.nY
    x = [ii*Δx for ii = 0:parms.nX-1]
    y = [ii*Δy for ii = 0:parms.nY-1]
    return x, y
end # generate_grid

function generate_initial_condition(parms::adv_diff_parameters, x::Vector{Float64}, y::Vector{Float64})
    # init
    init_sol = zeros(parms.ndof[1], parms.ndof[2])
    yT = transpose(y)
    # define
    if parms.initType == "gauss"
        init_sol[parms.dlo[1]:parms.dhi[1], parms.dlo[2]:parms.dhi[2]] = exp.(-200.0 .* ((x .- 0.25).^2 .+ (yT .- 0.25).^2))
    elseif  parms.initType == "square"
        init_sol[parms.dlo[1]:parms.dhi[1], parms.dlo[2]:parms.dhi[2]] = float.((x .> 0.2).*(x .< 0.3).*(yT .> 0.2).*(yT .< 0.3))
    elseif parms.initType == "cross"
        init_sol[parms.dlo[1]:parms.dhi[1], parms.dlo[2]:parms.dhi[2]] = 0.5 .* ( exp.(-200.0 .* (x .- 0.5).^2) .+ exp.(-200.0 .* (yT .- 0.5).^2))
    elseif parms.initType == "cross2"
        init_sol[parms.dlo[1]:parms.dhi[1], parms.dlo[2]:parms.dhi[2]] = max.(exp.(-200.0 .* (x .- 0.5).^2), exp.(-200.0 .* (yT .- 0.5).^2))
    else
        @error("unknown initType : $(parms.initType)")
    end
    #
    return init_sol
end # generate_initial_condition

function setup_coeffs(parms::adv_diff_parameters, x::Vector{Float64}, y::Vector{Float64})
    # allocate
    coeffs = zeros(parms.nX, parms.nY, 2*nHalo+1, 2)
    Δx, Δy = 1.0/parms.nX, 1.0/parms.nY
    yT = transpose(y)
    vX, vY = zeros(parms.nX, parms.nY), zeros(parms.nX, parms.nY)
    # hardcoded stencil parameters
    cAdv = reshape([ 1.0/12, -2.0/3,    0.0, 2.0/3, -1.0/12],(1,1,5))    # dim (nHalo*2+1)    
    cDif = reshape([-1.0/12,  4.0/3, -5.0/2, 4.0/3, -1.0/12],(1,1,5))    # dim (nHalo*2+1)
    # compute
    if parms.flowType == "diagonal"
            vX .= 1.0
            vY .= 1.0
    elseif parms.flowType == "circular"
        r = sqrt.( (x .- 0.5).^2 .+ (yT .- 0.5).^2 )
        phi= atan.(yT.-0.5, x .- 0.5)
        rho = exp.(-10.0 .* r.^2)
        vX = -r .* 2.0 .* π .* sin.(phi) .* rho
        vY =  r .* 2.0 .* π .* cos.(phi) .* rho
    elseif parms.flowType == "circular2"
        r = sqrt.( (x .- 0.5).^2 .+ (yT .- 0.5).^2 )
        phi= atan.(yT.-0.5, x .- 0.5)
        rho = exp.(-5.0 .* r.^2)
        vX = -r .* 2.0 .* π .* sin.(phi) .* sin.(4.0 * π .* r) .* rho
        vY =  r .* 2.0 .* π .* cos.(phi) .* sin.(4.0 * π .* r) .* rho
    else
        @warn("unknown flowType: $(parms.flowType), using vX, vY = 0.0")
        # do nothng already initialized to zero
    end
    coeffs[:, :, :, 1] = (-vX .* cAdv./Δx) .+ (parms.viscosity .* cDif./(Δx^2))
    coeffs[:, :, :, 2] = (-vY .* cAdv./Δy) .+ (parms.viscosity .* cDif./(Δy^2))
    #
    return coeffs
end # setup_coeffs

function update_RHS_and_boundary!(sol::Matrix{Float64}, tmp::Matrix{Float64}, RHS::Matrix{Float64}, coeffs::Array{Float64,4}, parms::adv_diff_parameters)
    # allocate
    nX, nY = parms.nX,parms.nY
    
    updateHalo!(sol)
    RHS .= 0.0
    #
    for ii in axes(coeffs, 3)
        # Derivative in X
        copyto!(tmp, CartesianIndices(tmp), sol, CartesianIndices((ii:nX+ii-1, parms.dlo[2]:parms.dhi[2])))
        tmp .*= coeffs[:, :, ii, 1]
        copyto!(RHS, CartesianIndices(tmp), RHS+tmp, CartesianIndices(tmp))
        
        # Derivative in Y
        copyto!(tmp, CartesianIndices(tmp), sol, CartesianIndices((parms.dlo[1]:parms.dhi[1], ii:nY+ii-1)))
        tmp .*= coeffs[:, :, ii, 2]
        #RHS += tmp
        copyto!(RHS, CartesianIndices(tmp), RHS+tmp, CartesianIndices(tmp))
    end
end # compute_RHS

function simulate_adv_diff(parms::adv_diff_parameters, u0::Matrix{Float64}, coeffs::Array{Float64,4})
    # allocate
    uEval = zeros(size(u0))
    tmp = zeros(parms.nX,parms.nY)
    u1 = zeros(parms.nX, parms.nY)
    copyto!(u1, CartesianIndices(u1), u0, CartesianIndices((parms.dlo[1]:parms.dhi[1], parms.dlo[2]:parms.dhi[2])))
    RHS = zeros(parms.nX, parms.nY)
    # time step
    Δt = parms.tEnd / parms.nSteps
    @btime simulation_loop!($parms, $u0, $tmp, $RHS, $uEval, $u1, $coeffs, $Δt) 
    #t1 = time()
    #simulation_loop!(parms, u0, RHS, uEval, u1, coeffs, Δt) 
    #tend = time() - t1
    #print("tWall: $(tend)\n")
    #print("tWall/DOF: $(tend/(parms.nSteps*parms.nX*parms.nX))\n")

end # simulate_adv_diff

function simulation_loop!(parms::adv_diff_parameters, u0::Matrix{Float64}, tmp::Matrix{Float64}, RHS::Matrix{Float64}, uEval::Matrix{Float64}, u1::Matrix{Float64}, coeffs::Array{Float64,4}, Δt::Float64)
    # just extra function for @btime
    for ti in 1:parms.nSteps    
        
        update_RHS_and_boundary!(u0, tmp, RHS, coeffs, parms)
        copyto!(uEval, CartesianIndices((parms.dlo[1]:parms.dhi[1], parms.dlo[2]:parms.dhi[2])), RHS, CartesianIndices((RHS)))
        RHS .*= Δt/6
        u1 += RHS
        
        
        uEval .*= Δt/2
        uEval += u0
        update_RHS_and_boundary!(uEval, tmp, RHS, coeffs, parms)
        copyto!(uEval, CartesianIndices((parms.dlo[1]:parms.dhi[1], parms.dlo[2]:parms.dhi[2])), RHS, CartesianIndices((RHS)))
        RHS .*= Δt/3
        u1 += RHS

        
        uEval .*= Δt/2
        uEval += u0
        update_RHS_and_boundary!(uEval, tmp, RHS, coeffs, parms)
        copyto!(uEval, CartesianIndices((parms.dlo[1]:parms.dhi[1], parms.dlo[2]:parms.dhi[2])), RHS, CartesianIndices((RHS)))
        RHS .*= Δt/3
        u1 += RHS

        
        uEval .*= Δt
        uEval += u0
        update_RHS_and_boundary!(uEval, tmp, RHS, coeffs, parms)
        copyto!(uEval, CartesianIndices((parms.dlo[1]:parms.dhi[1], parms.dlo[2]:parms.dhi[2])), RHS, CartesianIndices((RHS)))
        RHS .*= Δt/6
        u1 += RHS

        
        copyto!(u0, CartesianIndices((parms.dlo[1]:parms.dhi[1], parms.dlo[2]:parms.dhi[2])), u1, CartesianIndices((u1)))
        
    end
end # simulation_loop


## Parameters
const nHalo = 2
filename = "codes/01_advDiffSolver/input.txt"

# get parameters
parms = adv_diff_parameters()
read_inputs_and_update_parameters!(filename, parms)

# initialize
xvec, yvec = generate_grid(parms)
u0 = generate_initial_condition(parms, xvec, yvec)
coeffs = setup_coeffs(parms, xvec, yvec)

# plot initial condition 
fig1 = plot()
contourf!(fig1, xvec, yvec, u0[parms.dlo[1]:parms.dhi[1], parms.dlo[2]:parms.dhi[2]], color=:viridis, levels=20, aspect_ratio=:equal, size=(400,400))
plot!(fig1,xticks=0:0.2:1, yticks=0:0.2:1, xlims=(0.0,1.0), ylims=(0.0,1.0), title = "IC=$(parms.initType)", titlefontsize=10, label="")


# run simulation
simulate_adv_diff(parms, u0, coeffs)

# plot
fig2 = plot()
contourf!(fig2, xvec, yvec, u0[parms.dlo[1]:parms.dhi[1], parms.dlo[2]:parms.dhi[2]], color=:viridis, levels=20, aspect_ratio=:equal, size=(400,400))
plot!(fig2,xticks=0:0.2:1, yticks=0:0.2:1, xlims=(0.0,1.0), ylims=(0.0,1.0), title = "Final flow=$(parms.flowType)", titlefontsize=10, label="")


fig3 = plot(fig1,fig2,layout=(1,2),size=(800,400)) 